menu/src/main/java/com/jagrosh/jdautilities/menu/MessagePaginator.java

```diff
@@ -0,0 +1,579 @@
+         * with this Builder.
+         *
+         * @return The MessagePaginator built from this Builder.
+         *
+         * @throws java.lang.IllegalArgumentException
+         *         If one of the following is violated:
+         *         <ul>
+         *             <li>No {@link com.jagrosh.jdautilities.commons.waiter.EventWaiter EventWaiter} was set.</li>
+         *             <li>No function to paginate were set.</li>
+         *         </ul>
+         */
+        @Override
+        public MessagePaginator build()
+        {
+            Checks.check(waiter != null, "Must set an EventWaiter");
+            Checks.check(function != null, "Must include at least one item to paginate");
+
+            return new MessagePaginator(waiter, users, roles, timeout, unit, function, finalAction,
+                items, waitOnSinglePage,
+                bulkSkipNumber, wrapPageEnds, textToLeft, textToRight, allowTextInput);
+        }
+
+        /**
+         * Sets the {@link net.dv8tion.jda.core.entities.MessageEmbed MessageEmbed} to be displayed
+         * relative to the total page number and the current page as determined by the provided
+         * {@link java.util.function.BiFunction BiFunction}.
+         * <br>As the page changes, the BiFunction will re-process the current page number and the total
+         * page number, allowing for the displayed text of the Message to change depending on the page number.
+         *
+         * @param  embedFunction
+         *         The BiFunction that uses both current and total page numbers to get embed for the Message
+         *
+         * @return This builder
+         */
+        public Builder setEmbedFunction(BiFunction<Integer, Integer, MessageEmbed> embedFunction)
+        {
+            this.function = (pageNum, pages) -> new MessageBuilder().setEmbed(embedFunction.apply(pageNum, pages)).build();
+            return this;
+        }
+
+        /**
+         * Sets the {@link net.dv8tion.jda.core.entities.Message Message} to be displayed
+         * relative to the total page number and the current page as determined by the provided
+         * {@link java.util.function.BiFunction BiFunction}.
+         * <br>As the page changes, the BiFunction will re-process the current page number and the total
+         * page number, allowing for the displayed  Message to change depending on the page number.
+         *
+         * @param  messageFunction
+         *         The BiFunction that uses both current and total page numbers to get the Message
+         *
+         * @return This builder
+         */
+        public Builder setFunction(BiFunction<Integer, Integer, Message> messageFunction)
+        {
+            this.function = messageFunction;
+            return this;
+        }
+
+        /**
+         * Sets the {@link java.util.function.Consumer Consumer} to perform if the
+         * {@link com.jagrosh.jdautilities.menu.Paginator Paginator} times out.
```

Comment:
Documentation can be partially copied from existing menus of the same structure, *however* this and all other references to `Paginator` should be modified to respect the new entity.
IE: `Paginator` -> `MessagePaginator`

Reply:
oops I forgot that one